docs/config_file.rst
Configuration File
===================

contains configuration file details

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [ 0.7.1 ] - [ 2019-09-02 ]
### Added
-

### Changed
- re-factored folders


### Removed


### Fixed
- fixed issue with correct python library picking on tud win systems



## [ 0.7.0 ] - [ 2019-06-03 ]
### Added
- support for single qubit flux options (auto/manual modes)
- option to control generation of qasm files and dot graphs
- NPROCS=n variable can now be set for faster compilation to use n threads
- conda build recipe
- conda binary releases for Linux, Windows platform (not yet available for OSX due to a conda distribution issue)


### Changed
- openql is now public
- improved resource-constrained scheduling
- sweep point array is now optional
- support for barrier/wait on all qubits


### Removed
- set_sweep_points(sweep_points list, num of sweep points)


### Fixed
- resource-constrained qasm is generated by same scheduler for cc-light as is used to generate qisa
- Illegal parameter in gate_decomposition


## [ 0.6 ] - [ 2018-10-29 ]
### Added
- generated qasm code conforms to cQASM v1.0 specification
- added libqasm to pytest to test conformance of generated qasm


### Changed
- ALAP scheduler is the default option (Issue #193)
- compiling an empty program raises error (Issue #164)

### Removed


### Fixed
- tests are added to test option setting/getting (Issue #190)



## [ 0.5.5 ] - [ 2018-10-25 ]
### Added

### Changed
- simplified interface of Program.set_sweep_points (Issue #184)

### Removed

### Fixed
- instruction ordering to generate consistent qisa (Issue #190)
- stateful behaviour in OpenQL (Issue #171)


## [ 0.5.4 ] - [ 2018-10-17 ]
### Added

### Changed

### Removed

### Fixed
- qubit ordering in SMIS and SMIT instructions


## [ 0.5.3 ] - [ 2018-10-11 ]
### Added
- added detuning constraints for cclight

### Changed

### Removed

### Fixed
- alap scheduling for cclight


## [ 0.5.2 ] - [ 2018-10-10 ]
### Added

### Changed

### Removed

### Fixed
- wrong target qubits in the configuration files
- Jenkins test build profile to test against assembler


## [ 0.5.1 ] - [ 2018-09-12 ]
### Added
- API to obtain version number

### Changed

### Removed

### Fixed
- qisa format (removed comma)


## [ 0.5 ] - [ 2018-06-26 ]
### Added
- support for classical instructions
- support for flow control (selection and repetition)
- classical register manager implementation

### Changed
- measure instruction updated to support classical target register
- kernels are not any more fused to generate a single qisa program

### Removed
- kernel does not recieve iteration count, deprecated in favor of for-loop

### Fixed
- qisa format (pre-interval syntax updated)

## [ 0.4.1 ] - [ 2018-05-31 ]
### Added
-

### Changed
-

### Removed
- 

### Fixed
- getting started example


## [ 0.4 ] - [ 2018-05-19 ]
### Added
- kernel conjugation/un-compute feature
- multi-qubit control decomposition
- toffoli decomposition
- QASM loader for QASM v1.0 syntax check
- initial support for Quantumsim backend
- vebosity levels


### Changed
- program options can be set/get with simple api calls
- when adding gates, qubits should always be specified as list
- updated qisa-as support for tests

### Removed
- qisa-as is not a part of openql
- prog.compile() does not get optimiz/schedule/verbose options

### Fixed
- static iteration count for scheduled qasm
- roation angle printing


## [ 0.3 ] - [ 2017-10-24 ]
### Added
- CCLight eQASM compiler
- unittests using qisa-as
- simplified gate decompositions
- wait/barrier instructions


### Changed
-

### Removed
- 

### Fixed
- varying prepz duration
- M_PI issue in windows install


## [ 0.2 ] - [ 2017-08-18 ]
### Added
- CBox eQASM compiler
- Python and C++ interface
- Configuration file specifiction
- trace support for qumis code
- cmake based builds

### Changed
-

### Removed
- 

### Fixed
-


src/options.h
/**
 * @file   options.h
 * @date   05/2018
 * @author Imran Ashraf
 * @brief  Options class implementing parsing of supported options
 */

#ifndef QL_OPTIONS_H
#define QL_OPTIONS_H

#include <utils.h>
#include <CLI/CLI.hpp>
#include <iostream>

namespace ql
{
  class Options
  {
  private:
      CLI::App * app;
      std::map<std::string, std::string> opt_name2opt_val;

  public:
      Options(std::string app_name="testApp")
      {
          app = new CLI::App(app_name);

          // default values
          opt_name2opt_val["log_level"] = "LOG_NOTHING";
          opt_name2opt_val["output_dir"] = "test_output";
          opt_name2opt_val["unique_output"] = "no";
          opt_name2opt_val["optimize"] = "no";
          opt_name2opt_val["use_default_gates"] = "no";
          opt_name2opt_val["decompose_toffoli"] = "no";
          opt_name2opt_val["quantumsim"] = "no";

          opt_name2opt_val["scheduler"] = "ALAP";
          opt_name2opt_val["scheduler_uniform"] = "no";
          opt_name2opt_val["scheduler_commute"] = "yes";
          opt_name2opt_val["prescheduler"] = "yes";
          opt_name2opt_val["scheduler_post179"] = "yes";
          opt_name2opt_val["cz_mode"] = "manual";
          opt_name2opt_val["print_dot_graphs"] = "no";

          opt_name2opt_val["clifford_premapper"] = "yes";
          opt_name2opt_val["clifford_postmapper"] = "yes";

          opt_name2opt_val["mapper"] = "minextendrc";
          opt_name2opt_val["mapassumezeroinitstate"] = "no";
          opt_name2opt_val["mapinitone2one"] = "yes";
          opt_name2opt_val["mapprepinitsstate"] = "no";
          opt_name2opt_val["initialplace"] = "no";
          opt_name2opt_val["initialplace2qhorizon"] = "0";
          opt_name2opt_val["maplookahead"] = "noroutingfirst";
          opt_name2opt_val["mappathselect"] = "all";
          opt_name2opt_val["maprecNN2q"] = "no";
          opt_name2opt_val["mapselectmaxlevel"] = "0";
          opt_name2opt_val["mapselectmaxwidth"] = "min";
          opt_name2opt_val["mapselectswaps"] = "all";
          opt_name2opt_val["maptiebreak"] = "random";
          opt_name2opt_val["mapusemoves"] = "yes";
          opt_name2opt_val["mapreverseswap"] = "yes";

          opt_name2opt_val["write_qasm_files"] = "no";
          opt_name2opt_val["write_report_files"] = "no";

          // add options with default values and list of possible values
          app->add_set_ignore_case("--log_level", opt_name2opt_val["log_level"], 
            {"LOG_NOTHING", "LOG_CRITICAL", "LOG_ERROR", "LOG_WARNING", "LOG_INFO", "LOG_DEBUG"}, "Log levels", true);
          app->add_option("--output_dir", opt_name2opt_val["output_dir"], "Name of output directory", true);
          app->add_set_ignore_case("--unique_output", opt_name2opt_val["unique_output"], {"no", "yes"}, "Make output files unique", true);
          app->add_set_ignore_case("--prescheduler", opt_name2opt_val["prescheduler"], {"no", "yes"}, "Run qasm (first) scheduler?", true);
          app->add_set_ignore_case("--scheduler_post179", opt_name2opt_val["scheduler_post179"], {"no", "yes"}, "Issue 179 solution included", true);
          app->add_set_ignore_case("--print_dot_graphs", opt_name2opt_val["print_dot_graphs"], {"no", "yes"}, "Print (un-)scheduled graphs in DOT format", true);
          app->add_set_ignore_case("--scheduler", opt_name2opt_val["scheduler"], {"ASAP", "ALAP"}, "scheduler type", true);
          app->add_set_ignore_case("--scheduler_uniform", opt_name2opt_val["scheduler_uniform"], {"yes", "no"}, "Do uniform scheduling or not", true);
          app->add_set_ignore_case("--scheduler_commute", opt_name2opt_val["scheduler_commute"], {"yes", "no"}, "Commute gates when possible, or not", true);
          app->add_set_ignore_case("--use_default_gates", opt_name2opt_val["use_default_gates"], {"yes", "no"}, "Use default gates or not", true);
          app->add_set_ignore_case("--optimize", opt_name2opt_val["optimize"], {"yes", "no"}, "optimize or not", true);
          app->add_set_ignore_case("--clifford_premapper", opt_name2opt_val["clifford_premapper"], {"yes", "no"}, "clifford optimize before mapping yes or not", true);
          app->add_set_ignore_case("--clifford_postmapper", opt_name2opt_val["clifford_postmapper"], {"yes", "no"}, "clifford optimize after mapping yes or not", true);
          app->add_set_ignore_case("--decompose_toffoli", opt_name2opt_val["decompose_toffoli"], {"no", "NC", "MA"}, "Type of decomposition used for toffoli", true);
          app->add_set_ignore_case("--quantumsim", opt_name2opt_val["quantumsim"], {"no", "yes", "qsoverlay"}, "Produce quantumsim output, and of which kind", true);
          app->add_set_ignore_case("--cz_mode", opt_name2opt_val["cz_mode"], {"manual", "auto"}, "CZ mode", true);

          app->add_set_ignore_case("--mapper", opt_name2opt_val["mapper"], {"no", "base", "baserc", "minextend", "minextendrc", "maxfidelity", "maxfidelity_debug"}, "Mapper heuristic", true);
          app->add_set_ignore_case("--mapinitone2one", opt_name2opt_val["mapinitone2one"], {"no", "yes"}, "Initialize mapping of virtual qubits one to one to real qubits", true);
          app->add_set_ignore_case("--mapprepinitsstate", opt_name2opt_val["mapprepinitsstate"], {"no", "yes"}, "Prep gate leaves qubit in zero state", true);
          app->add_set_ignore_case("--mapassumezeroinitstate", opt_name2opt_val["assumezeroinitstate"], {"no", "yes"}, "Assume that qubits are initialized to zero state", true);
          app->add_set_ignore_case("--initialplace", opt_name2opt_val["initialplace"], {"no","yes","1s","10s","1m","10m","1h","1sx","10sx","1mx","10mx","1hx"}, "Initialplace qubits before mapping", true);
          app->add_set_ignore_case("--initialplace2qhorizon", opt_name2opt_val["initialplace2qhorizon"], {"0","1","2","3","4","5","6","7","8","9", "10","11","12","13","14","15","16","17","18","19","20","30","40","50","60","70","80","90","100"}, "Initialplace considers only this number of initial two-qubit gates", true);
          app->add_set_ignore_case("--maplookahead", opt_name2opt_val["maplookahead"], {"no", "1qfirst", "noroutingfirst", "all"}, "Strategy wrt selecting next gate(s) to map", true);
          app->add_set_ignore_case("--mappathselect", opt_name2opt_val["mappathselect"], {"all", "borders"}, "Which paths: all or borders", true);
          app->add_set_ignore_case("--mapselectswaps", opt_name2opt_val["mapselectswaps"], {"one", "all", "earliest"}, "Select only one swap, or earliest, or all swaps for one alternative", true);
          app->add_set_ignore_case("--maprecNN2q", opt_name2opt_val["maprecNN2q"], {"no","yes"}, "Recursing also on NN 2q gate?", true);
          app->add_set_ignore_case("--mapselectmaxlevel", opt_name2opt_val["mapselectmaxlevel"], {"0","1","2","3","4","5","6","7","8","9","10","inf"}, "Maximum recursion in selecting alternatives on minimum extension", true);
          app->add_set_ignore_case("--mapselectmaxwidth", opt_name2opt_val["mapselectmaxwidth"], {"min","minplusone","minplushalfmin","minplusmin","all"}, "Maximum width number of alternatives to enter recursion with", true);
          app->add_set_ignore_case("--maptiebreak", opt_name2opt_val["maptiebreak"], {"first", "last", "random", "critical"}, "Tie break method", true);
          app->add_set_ignore_case("--mapusemoves", opt_name2opt_val["mapusemoves"], {"no", "yes", "0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20"}, "Use unused qubit to move thru", true);
          app->add_set_ignore_case("--mapreverseswap", opt_name2opt_val["mapreverseswap"], {"no", "yes"}, "Reverse swap operands when better", true);

          app->add_set_ignore_case("--write_qasm_files", opt_name2opt_val["write_qasm_files"], {"yes", "no"}, "write (un-)scheduled (with and without resource-constraint) qasm files", true);
          app->add_set_ignore_case("--write_report_files", opt_name2opt_val["write_report_files"], {"yes", "no"}, "write report files on circuit characteristics and pass results", true);
      }

      void print_current_values()
      {
          std::cout << "log_level: " << opt_name2opt_val["log_level"] << std::endl
                    << "output_dir: " << opt_name2opt_val["output_dir"] << std::endl
                    << "unique_output: " << opt_name2opt_val["unique_output"] << std::endl
                    << "optimize: " << opt_name2opt_val["optimize"] << std::endl
                    << "use_default_gates: " << opt_name2opt_val["use_default_gates"] << std::endl
                    << "decompose_toffoli: " << opt_name2opt_val["decompose_toffoli"] << std::endl
                    << "quantumsim: " << opt_name2opt_val["quantumsim"] << std::endl
                    << "prescheduler: " << opt_name2opt_val["prescheduler"] << std::endl
                    << "scheduler: " << opt_name2opt_val["scheduler"] << std::endl
                    << "scheduler_uniform: " << opt_name2opt_val["scheduler_uniform"] << std::endl
                    << "clifford_premapper: " << opt_name2opt_val["clifford_premapper"] << std::endl
                    << "mapper: "           << opt_name2opt_val["mapper"] << std::endl
                    << "mapinitone2one: "   << opt_name2opt_val["mapinitone2one"] << std::endl
                    << "initialplace: "     << opt_name2opt_val["initialplace"] << std::endl
                    << "initialplace2qhorizon: "<< opt_name2opt_val["initialplace2qhorizon"] << std::endl
                    << "maplookahead: "     << opt_name2opt_val["maplookahead"] << std::endl
                    << "mappathselect: "    << opt_name2opt_val["mappathselect"] << std::endl
                    << "maptiebreak: "      << opt_name2opt_val["maptiebreak"] << std::endl
                    << "mapusemoves: "      << opt_name2opt_val["mapusemoves"] << std::endl
                    << "mapreverseswap: "   << opt_name2opt_val["mapreverseswap"] << std::endl
                    << "mapselectswaps: "   << opt_name2opt_val["mapselectswaps"] << std::endl
                    << "clifford_postmapper: " << opt_name2opt_val["clifford_postmapper"] << std::endl
                    << "scheduler_post179: " << opt_name2opt_val["scheduler_post179"] << std::endl
                    << "scheduler_commute: " << opt_name2opt_val["scheduler_commute"] << std::endl
                    << "cz_mode: " << opt_name2opt_val["cz_mode"] << std::endl
                    << "print_dot_graphs: " << opt_name2opt_val["print_dot_graphs"] << std::endl;
      }

      void help()
      {
          std::cout << app->help() << std::endl;
      }

      void set(std::string opt_name, std::string opt_value)
      {
          try
          {
            std::vector<std::string> opts = {opt_value, "--"+opt_name};
            app->parse(opts);
          }
          catch (const std::exception &e)
          {
            app->reset();
            EOUT("Un-known option:"<< e.what());
            throw ql::exception("Error parsing options. "+std::string(e.what())+" !",false);
          }
          app->reset();
      }

      std::string get(std::string opt_name)
      {
        std::string opt_value("UNKNOWN");
        if( opt_name2opt_val.find(opt_name) != opt_name2opt_val.end() )
        {
            opt_value = opt_name2opt_val[opt_name];
        }
        else
        {
            EOUT("Un-known option:"<< opt_name);
        }
        return opt_value;
      }
  };

  namespace options
  {
      static ql::Options ql_options("OpenQL Options");
      void print()
      {
          ql_options.help();
      }
      void print_current_values()
      {
          ql_options.print_current_values();
      }
      void set(std::string opt_name, std::string opt_value)
      {
          ql_options.set(opt_name, opt_value);

          if(opt_name == "log_level")
              ql::utils::logger::set_log_level(opt_value);
          else if(opt_name == "output_dir")
              ql::utils::make_output_dir(opt_value);
      }
      std::string get(std::string opt_name)
      {
          return ql_options.get(opt_name);
      }
  } // namespace option
} // namespace ql

#endif

/*        
        void set(std::string opt_name, std::string opt_value)
        {
            if( _options.find(opt_name) != _options.end() )
            {
                if(opt_name == "log_level")
                    ql::utils::logger::set_log_level(opt_value);
                else if(opt_name == "output_dir")
                    ql::utils::make_output_dir(opt_value);
                else if(opt_name == "scheduler" && opt_value == "ASAP" && opt_value == "ALAP")
                {

                }

                _options[opt_name] = opt_value;
            }
            else
            {
                std::cerr << "[OPENQL] " << __FILE__ <<":"<< __LINE__ 
                          <<" Error: Un-known option:"<< opt_name << std::endl;
            }
        }

        void print()
        {
            std::cout << "OpenQL Options:\n";
            for(auto elem : _options)
            {
                std::cout << "    " << elem.first << " : " << elem.second << "\n";
            }            
        }
*/

